WordPress/gutenberg
# useFocusOutside: rewrite hook to TypeScript, rewrite tests to model RTL and user-event

## What?


While trying to fix #40912 I got to rewrite the hook to TypeScript and the tests to modern RTL techniques (including `user-event`)

## Why?


Better tests and TypeScript as a more solid foundation for working on a tricky problem (which I haven't solved, but I thought I'd push these improvements anyway)

## How?


Tests:
- using `user-event` instead of `fireEvent`(or even imperative events called on each element) where possible
- using jest mocking APIs for `document.hasFocus`
- removed useless `TestComponent` wrapper

Hook:
- no runtime changes
- used more specific types for event listeners (e.g. `FocusEventHandler` etc)
- updated the exported type in `useDialog`

## Testing Instructions





Check the `Modal` Storybook example (and modals around the codebase) and make sure that they behave as currently on `trunk`

Code changes can reviewed more easily when approached commit-by-commit
